Adding information about secrets in OCI Storage (#259)
rancher/fleet-docs: docs/oci-storage.md

* **GitRepo-level secret:** A custom secret for specific GitRepo resouces.
  * This is a user-defined secret can have any name and must be referenced in the `GitRepo` resource. 

### GitRepo Configuration Example

If a secret named `ocistorage` exists in a namespace. Fleet automatically uses it as the OCI registry for all GitRepo resources in that namespace, and you don’t need to add `ociRegistrySecret` in every GitRepo.

You can explicitly specify a different secret in the GitRepo. The GitRepo spec is extended with information about the OCI Registry. You have to update your GitRepo resource to include a reference to the secret:

```yaml
metadata:
  name: coffee-shop-deployments
  namespace: fleet-local
spec:
  repo: https://github.com/example/coffee-shop-fleet
  branch: main
  paths:
    - ./manifests
  ociRegistrySecret: ocistorage
```

The `ociRegistrySecret` references a secret in the same namespace, which contains the address and credentials needed to read/write to the OCI registry. 

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: ocistorage
  namespace: fleet-local

> type: fleet.cattle.io/bundle-oci-storage/v1alpha1

Reviewer: We'd need to be more explicit saying that the type in **enforced** to be `fleet.cattle.io/bundle-oci-storage/v1alpha1` otherwise the secret will be rejected by Fleet

data:
  reference: <base64-encoded-registry-url>
  username: <base64-encoded-user>
  password: <base64-encoded-password>
  insecureSkipTLS: <base64-encoded-true/false>
  basicHTTP: <base64-encoded-true/false>
# Optional
  agentUsername: <base64-encoded-readonly-user>
  agentPassword: <base64-encoded-password>
```
:::note
Only the reference field is required. All other fields are optional. 
:::

If no username or password is specified, Fleet accesses the registry without authentication. You should follow these best practices:

* Use read-only `agentUsername` and `agentPassword` credentials for agents to enhance security.  
  * However, if you don’t set these credentials, the agent uses user credentials with read/write permissions.  

> * Insecure TLS defaults to `false`, and basic http is disabled by default.   

Reviewer: ```suggestion
* Insecure TLS defaults to `false`, and basic HTTP is disabled by default.   
```

  * Fleet allows these flags for development and testing purposes, but they should never be used in production.  
  * If you use these tags, you expose your app to security vulnerabilities, and put cluster workloads, and credentials at a risk of tampering.  

> * Changing the secret does not update the deployment. The new storage registry is used only after the next Git update or when you trigger a force update.

Reviewer: This is how Fleet works, right?
Not an item of best practices.... I guess it should go in a different place

:::note
Fleet checks for the integrity of OCI artifacts and Fleet tags OCI artifact as Latest
:::

### Fleet Example

> Consider the GitRepo file for fleet-examples repository. 

Reviewer: ```suggestion
Consider the GitRepo file from [fleet-examples](https://github.com/rancher/fleet-examples) repository. 
```

Is it really a file existing in the fleet-examples repo?
If it does not exist in the repository I would not mention it at all. (No need to create it in fleet-examples)

```yaml
apiVersion: fleet.cattle.io/v1alpha1
kind: GitRepo
metadata:
  name: frontend-oci
  namespace: fleet-local
spec:
  repo: https://github.com/your-org/fleet-oci-example.git
  branch: main
  paths:
    - ./frontend
  ociRegistrySecret: ocistorage
```

> To define OCI Secret, you can create an oci-secret.yaml file

Reviewer: ```suggestion
To define the OCI Secret, you can create and apply a yaml file like the following one:
```

I would not mention a specific `yaml` name, because it might sound as we need that specific name for it (it can be any name)

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: ocistorage
  namespace: fleet-local
type: fleet.cattle.io/bundle-oci-storage/v1alpha1
data:
  username: <base64-encoded-user>
  password: <base64-encoded-password>
  insecureSkipTLS: <base64-encoded-true/false>
  basicHTTP: <base64-encoded-true/false>
  agentUsername: <base64-encoded-readonly-user>
  agentPassword: <base64-encoded-password>
```

Run `kubectl apply -f secrets/oci-secret.yaml` before applying the GitRepo.

Or you can use `kubectl` command to create the `ocistorage` secret using unencoded text. Kubernetes converts them to base64 encoded for storing the secret.

> 

Reviewer: I would mention that the `reference` field is a URL without prefixes. 
For example, no `oci://` prefix should be added.

# OCI Storage

> Fleet stores Kubernetes bundle resources in etcd twice by default. This is done using the K8s API. However, etcd has storage limits and is not optimized for large workloads. As your deployments scale, it is more efficient to store these resources externally in an OCI (Open Container Initiative) registry.

Reviewer: I think mentioning that Fleet stores resources twice is an implementation detail knowledge not strictly needed by the final user.

:::note
Fleet recommends you to compress and encode the bundle into base64 to reduce the resource size.
:::

> If your bundle resource is greater than X MB(size), then it would be a better alternative to switch to OCI Registry as Storage. 

Reviewer: How much is X?
I would, maybe, mention X as the limits of `etcd` configured for that particular cluster..

Using an OCI registry helps you:

* Reduce etcd load by offloading large bundle content.  
* Use a standardized storage backend for large manifests or Helm charts.

Once the OCI registry is enabled, Fleet will use it as the source for storing Bundle resources. When Fleet can't access the OCI registry, it won't fall back to default etcd storage. Instead, it logs errors.

![A visual asset displaying the flow of Fleet with OCI Storage.](../static/img/fleet-ociStorage-flow.png)

## **Prerequisites**

* A running OCI registry.  
* A Kubernetes secret with valid credentials.  
* A Fleet installation (v2.12.0 or later) .

## Creating an OCI Storage Secret

To enable OCI Storage, you have to define a specific kind of secret. There are two ways of defining secrets

* **Global secret (ocistorage)** : Applies to all GitRepos in the namespace.  

>   * This is the fallback secret. If no GitRepo-level secret is specified, Fleet uses this secret for all GitRepos in the namespace.  

Reviewer: I think we need to mention explicitly that the name of the namespace secret is `ocistorage`. I see you are mentioning the name but I think we have to be more explicit.

* **GitRepo-level secret (ociRegistrySecret):** – Overrides the global secret for a specific GitRepo.  
  * This is a user-defined secret that you reference in the `GitRepo` resource. 

> 

Reviewer: I would also mention that the name is not restricted to anything in this case

:::note
If the referenced secret doesn't exist or contains invalid credentials, Fleet logs an error and skips deployment. It doesn't fall back to etcd.
:::

### GitRepo Configuration Example

> The GitRepo spec is extended with information about the OCI Registry. You must update your GitRepo resource to include a reference to the secret:

Reviewer: This is not strictly required when using the global (namespace) secret.
When that secret exists all `GitRepo`s defined for that namespace will use the OCI registry defined.

```yaml
metadata:
  name: coffee-shop-deployments
  namespace: fleet-local
spec:
  repo: https://github.com/example/coffee-shop-fleet
  branch: main
  paths:
    - ./manifests
  ociRegistrySecret: ocistorage
```

The `ociRegistrySecret` references a secret in the same namespace, which contains the address and credentials needed to read/write to the OCI registry. 

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: ocistorage
  namespace: fleet-local
type: fleet.cattle.io/bundle-oci-storage/v1alpha1
data:
  reference: <base64-encoded-registry-url>
  username: <base64-encoded-user>
  password: <base64-encoded-password>
  insecureSkipTLS: <base64-encoded-true/false>
  basicHTTP: <base64-encoded-true/false>
# Optional
  agentUsername: <base64-encoded-readonly-user>
  agentPassword: <base64-encoded-password>
```
:::note

> If your user is not defined in the spec of the Repo.Then Fleet tries to use default credentials.

Reviewer: What does this mean? 
Default credentials means no username/password? 

I would also mention that only the `reference` field is required, the rest are optional.
If no username/password are defined then no authentication will be used to access the registry.

:::

You should follow these best practices:

* Use read-only `agentUsername` and `agentPassword` credentials for agents to enhance security.  

>   * However, if you don’t send these credentials, the agent uses user credentials with read/write permissions.  

Reviewer: ```suggestion
  * However, if you don’t set these credentials, the agent uses user credentials with read/write permissions.  
```

> * Insecure TLS defaults to `false`, and http is disabled by default.   

Reviewer: ```suggestion
* Insecure TLS defaults to `false`, and basic http is disabled by default.   
```

  * Fleet allows these flags for development and testing purposes, but they should never be used in production.  

>   * If you use these tags, you expose your app to security vulnerabilities, and put cluster workloads, and credentials at a risk of tampering with.  

Reviewer: ```suggestion
  * If you use these tags, you expose your app to security vulnerabilities, and put cluster workloads, and credentials at a risk of tampering.  
```

* Changing the secret does not update the deployments, only on the next GIT update it would use the new storage registry.

> 

Reviewer: Or ForceUpdate

> Fleet uses `gitcredential` secret when the `clientSecretName` field is empty in a GitRepo. It uses global `ocistorage` as a fallback if `ociRegistrySecret` is not defined.

Reviewer: Why are we talking about `gitcredential` and `clientSecretName`? 

I think we need to better explain how the secrets work. Because the secrets enable the usage of the OCI registry.

:::note
Fleet checks for the integrity of OCI artifacts and Fleet tags OCI artifact as Latest
:::

### Fleet Example

Consider the GitRepo file for fleet-examples repository. 

```yaml
apiVersion: fleet.cattle.io/v1alpha1
kind: GitRepo
metadata:
 name: frontend-oci
 namespace: fleet-local
spec:
 repo: https://github.com/your-org/fleet-oci-example.git
 branch: main
 paths:
 - ./frontend/frontend-deployment.yaml

>     - ./frontend/frontend-service.yaml

Reviewer: Bad indentation.
I also think it will be enough with:
```yaml
  paths:
    - ./frontend
```
Paths are folders.

  ociRegistrySecret: ocistorage
```

> To define OCI Secret, you must create an oci-secret.yaml file

Reviewer: This is not completely true.
You also mention that you can also create the secret using `kubectl`.
So I would no say that you `must` create the yaml... you have the option to create the yaml or not.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: ocistorage
  namespace: fleet-local
type: fleet.cattle.io/bundle-oci-storage/v1alpha1
data:
  username: <base64-encoded-user>
  password: <base64-encoded-password>
  insecureSkipTLS: <base64-encoded-true/false>
  basicHTTP: <base64-encoded-true/false>
  agentUsername: <base64-encoded-readonly-user>
  agentPassword: <base64-encoded-password>
  
```

Run `kubectl apply -f secrets/oci-secret.yaml` before applying the GitRepo.

Or you can use `kubectl` command to create the `ocistorage` secret using unencoded text. Kubernetes converts them to base64 encoded for storing the secret.

```bash
kubectl -n fleet-local create secret generic ocistorage \
  --type=fleet.cattle.io/bundle-oci-storage/v1alpha1 \
  --from-literal=username=fleet-ci \
  --from-literal=password=fleetRocks \
  --from-literal=reference=192.168.1.39:8082 \
  --from-literal=insecureSkipTLS=true \
  --from-literal=basicHTTP=false \
  --from-literal=agentUsername=fleet-ci-readonly \
  --from-literal=agentPassword=readonlypass
```

To validate your secret, you can run:

> `kubectl get secret ocistorage -n fleet-local -o yaml`

Reviewer: I would also add the command to show the values of the secret decoded.
In fact I see you list the command in the screenshot.

Validating the secret is, anyway, part of kubernetes itself, it's not tied to Fleet.

You can explicitly specify a different secret in the `GitRepo`. The `GitRepo` spec is extended with information about the OCI Registry. You have to update your `GitRepo` resource to include a reference to the secret:

```yaml
metadata:
  name: coffee-shop-deployments
  namespace: fleet-local
spec:
  repo: https://github.com/example/coffee-shop-fleet
  branch: main
  paths:
    - ./manifests
  ociRegistrySecret: ocistorage
```

The `ociRegistrySecret` references a secret in the same namespace, which contains the address and credentials needed to read/write to the OCI registry. 

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: ocistorage
  namespace: fleet-local
type: fleet.cattle.io/bundle-oci-storage/v1alpha1
data:
  reference: <base64-encoded-registry-url> # Only the reference field is required. All other fields are optional. 
  username: <base64-encoded-user>
  password: <base64-encoded-password>
  insecureSkipTLS: <base64-encoded-true/false>
  basicHTTP: <base64-encoded-true/false>
  agentUsername: <base64-encoded-readonly-user>
  agentPassword: <base64-encoded-password>
```
:::note

> The secret must have type: `fleet.cattle.io/bundle-oci-storage/v1alpha1`. Fleet requires this value and rejects any secret that with a different type.

Reviewer: ```suggestion
The secret must have the type: `fleet.cattle.io/bundle-oci-storage/v1alpha1`. Fleet requires this value and rejects any secret with a different type.
```
